rust/tests/datafusion_test.rs

```diff
@@ -4,12 +4,95 @@ mod s3_common;
 
 #[cfg(feature = "datafusion-ext")]
 mod datafusion {
-    use std::sync::Arc;
+    use std::{collections::HashSet, sync::Arc};
 
-    use arrow::array::*;
-    use datafusion::error::Result;
-    use datafusion::execution::context::SessionContext;
+    use arrow::{
+        array::*,
+        datatypes::{
+            DataType as ArrowDataType, Field as ArrowField, Schema as ArrowSchema,
+            SchemaRef as ArrowSchemaRef,
+        },
+        record_batch::RecordBatch,
+    };
+    use datafusion::datasource::TableProvider;
+    use datafusion::error::{DataFusionError, Result};
+    use datafusion::execution::context::{SessionContext, TaskContext};
+    use datafusion::logical_expr::Expr;
+    use datafusion::logical_plan::Column;
+    use datafusion::physical_plan::{
+        coalesce_partitions::CoalescePartitionsExec, common, file_format::ParquetExec,
+        metrics::Label, visit_execution_plan, ExecutionPlan, ExecutionPlanVisitor,
+    };
     use datafusion::scalar::ScalarValue;
+    use deltalake::{action::SaveMode, operations::DeltaCommands, DeltaTable, DeltaTableMetaData};
+    use std::collections::HashMap;
+
+    fn get_scanned_files(node: &dyn ExecutionPlan) -> HashSet<Label> {
+        node.metrics()
+            .unwrap()
+            .clone()
+            .iter()
+            .cloned()
+            .flat_map(|m| m.labels().to_vec())
+            .collect()
+    }
+
+    #[derive(Debug, Default)]
+    pub struct ExecutionMetricsCollector {
+        scanned_files: HashSet<Label>,
+    }
+
+    impl ExecutionMetricsCollector {
+        fn num_scanned_files(&self) -> usize {
+            self.scanned_files.len()
+        }
+    }
+
+    impl ExecutionPlanVisitor for ExecutionMetricsCollector {
+        type Error = DataFusionError;
+
+        fn pre_visit(
+            &mut self,
+            plan: &dyn ExecutionPlan,
+        ) -> std::result::Result<bool, Self::Error> {
+            if let Some(exec) = plan.as_any().downcast_ref::<ParquetExec>() {
+                let files = get_scanned_files(exec);
+                self.scanned_files.extend(files);
+            }
+            Ok(true)
+        }
+    }
```

Comment:
Just did this for testing right now, but I would like to use something like this to collect the statistics we need for proper conflict resolution.